// fapi-client/JS/src/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The base URL that all requests are sent to.
 */
exports.Base = 'https://fapi.wrmsr.io';
/**
*  All documented fAPI endpoints.
*/
var Routes;
(function (Routes) {
    Routes["_4CHAN"] = "/4chan";
    Routes["_9GAG"] = "/9gag";
    Routes["ADIDAS"] = "/adidas";
    Routes["ADW"] = "/adminwalk";
    Routes["AIMAGIK"] = "/ai_magik";
    Routes["AJIT"] = "/ajitpai";
    Routes["AMERICA"] = "/america";
    Routes["ANALYSIS"] = "/analysis";
    Routes["AUSTIN"] = "/austin";
    Routes["AUTISM"] = "/autism";
    Routes["BANDICAM"] = "/bandicam";
    Routes["BERNIE"] = "/bernie";
    Routes["BINOCULARS"] = "/binoculars";
    Routes["BLACKIFY"] = "/blackify";
    Routes["BLACKPANTHER"] = "/blackpanther";
    Routes["BOBROSS"] = "/bobross";
    Routes["BUZZFEED"] = "/buzzfeed";
    Routes["CHANGEMYMIND"] = "/changemymind";
    Routes["COMPOSITE"] = "/composite";
    Routes["CONSENT"] = "/consent";
    Routes["COOLGUY"] = "/coolguy";
    Routes["DAYS"] = "/days";
    Routes["DEEPFRY"] = "/deepfry";
    Routes["DEPRESSION"] = "/depression";
    Routes["DISABLED"] = "/disabled";
    Routes["DORK"] = "/dork";
    Routes["DUCKDUCKGO"] = "/duckduckgo";
    Routes["DUCKDUCKGOIMAGES"] = "/duckduckgoimages";
    Routes["EDGES"] = "/edges";
    Routes["EDGES2EMOJIS"] = "/edges2emojis";
    Routes["EDGES2EMOJISGIF"] = "/gifedges2emojis";
    Routes["EDGES2PORN"] = "/edges2porn";
    Routes["EDGES2PORNGIF"] = "/gifedges2porn";
    Routes["EMOJIFY"] = "/emojify";
    Routes["EMOJIMOSAIC"] = "/emojimosaic";
    Routes["EVAL"] = "/eval";
    Routes["EVALMAGIK"] = "/evalmagik";
    Routes["EXCUSE"] = "/excuse";
    Routes["EYES"] = "/eyes";
    Routes["FACEDETECTION"] = "/detect_faces";
    Routes["FACEMAGIK"] = "/facemagik";
    Routes["FACEOVERLAY"] = "/faceoverlay";
    Routes["FACESWAP"] = "/faceswap";
    Routes["GABEN"] = "/gaben";
    Routes["GAY"] = "/gay";
    Routes["GLITCH"] = "/glitch";
    Routes["GLOW"] = "/glow";
    Routes["GOD"] = "/god";
    Routes["GOLDSTAR"] = "/goldstar";
    Routes["GRILL"] = "/grill";
    Routes["HACKER"] = "/hacker";
    Routes["HAWKING"] = "/hawking";
    Routes["HYPERCAM"] = "/hypercam";
    Routes["IDUBBBZ"] = "/idubbbz";
    Routes["IFUNNY"] = "/ifunny";
    Routes["IMAGESCRIPT"] = "/image_script";
    Routes["IMAGETAGPARSER"] = "/parse_tag";
    Routes["ISRAEL"] = "/israel";
    Routes["JACK"] = "/jack";
    Routes["JACKOFF"] = "/jackoff";
    Routes["JESUS"] = "/jesus";
    Routes["KEEMSTAR"] = "/keemstar";
    Routes["KEEMSTAR2"] = "/keemstar2";
    Routes["KEKISTAN"] = "/kekistan";
    Routes["KIRBY"] = "/kirby";
    Routes["LEGO"] = "/lego";
    Routes["LINUS"] = "/linus";
    Routes["LOGAN"] = "/logan";
    Routes["LOGOUT"] = "/logout";
    Routes["MAGIKSCRIPT"] = "/magikscript";
    Routes["MEMORIAL"] = "/memorial";
    Routes["MIRANDA"] = "/miranda";
    Routes["MISTAKE"] = "/mistake";
    Routes["NOOSEGUY"] = "/nooseguy";
    Routes["NORTHKOREA"] = "/northkorea";
    Routes["OLDGUY"] = "/oldguy";
    Routes["OWO"] = "/owo";
    Routes["PATHLIST"] = "/pathlist";
    Routes["PERFECTION"] = "/perfection";
    Routes["PISTOL"] = "/pistol";
    Routes["PIXELATE"] = "/pixelate";
    Routes["PNE"] = "/pne";
    Routes["PORNHUB"] = "/pornhub";
    Routes["PORTAL"] = "/portal";
    Routes["PRESIDENTIAL"] = "/presidential";
    Routes["PROXY"] = "/proxy";
    Routes["QUOTE"] = "/quote";
    Routes["RACECARD"] = "/racecard";
    Routes["REALFACT"] = "/realfact";
    Routes["RECAPTCHA"] = "/recaptcha";
    Routes["REMINDER"] = "/reminder";
    Routes["RESIZE"] = "/resize";
    Routes["RESPECTS"] = "/respects";
    Routes["RETRO"] = "/retro";
    Routes["REXTESTER"] = "/rextester";
    Routes["RTX"] = "/rtx";
    Routes["RUSSIA"] = "/russia";
    Routes["SCREENSHOT"] = "/screenshot";
    Routes["SHIT"] = "/shit";
    Routes["SHOOTING"] = "/shooting";
    Routes["SHOTGUN"] = "/shotgun";
    Routes["SIMPSONSDISABLED"] = "/simpsons_disabled";
    Routes["SMG"] = "/smg";
    Routes["SNAPCHAT"] = "/snapchat";
    Routes["SONIC"] = "/sonic";
    Routes["SPAIN"] = "/spain";
    Routes["STARMAN"] = "/starman";
    Routes["STEAMPLAYING"] = "/steamplaying";
    Routes["STOCK"] = "/stock";
    Routes["SUPREME"] = "/supreme";
    Routes["THINKING"] = "/thinking";
    Routes["THONKIFY"] = "/thonkify";
    Routes["TRANS"] = "/trans";
    Routes["TRUMP"] = "/trump";
    Routes["UGLY"] = "/ugly";
    Routes["UK"] = "/uk";
    Routes["UNMAGIK"] = "/unmagik";
    Routes["URBANDICTIONARY"] = "/urbandictionary";
    Routes["URLIFY"] = "/urlify";
    Routes["USSR"] = "/ussr";
    Routes["VENDING"] = "/vending";
    Routes["WATCHMOJO"] = "/watchmojo";
    Routes["WHEEZE"] = "/wheeze";
    Routes["WIKIHOW"] = "/wikihow";
    Routes["WONKA"] = "/wonka";
    Routes["WTH"] = "/worsethanhitler";
    Routes["YUSUKE"] = "/yusuke";
    Routes["ZOOM"] = "/zoom";
    Routes["ZUCKERBERG"] = "/zuckerberg";
})(Routes = exports.Routes || (exports.Routes = {}));
