// fapi-client/JS/src/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
* All HTTP request methods the client may use.
*/
var HttpMethods;
(function (HttpMethods) {
    HttpMethods["GET"] = "GET";
    HttpMethods["POST"] = "POST";
})(HttpMethods = exports.HttpMethods || (exports.HttpMethods = {}));
/**
* Each endpoint either returns JSON or string/buffer. The request function takes one of these as an option.
*/
var ReturnMethods;
(function (ReturnMethods) {
    ReturnMethods["JSON"] = "json";
    ReturnMethods["BODY"] = "body";
})(ReturnMethods = exports.ReturnMethods || (exports.ReturnMethods = {}));
/**
 * All specific headers that may be handled within the client itself.
 */
var ReturnHeaders;
(function (ReturnHeaders) {
    ReturnHeaders["IMAGESCRIPT_CPUTIME"] = "x-imagescript-cputime";
    ReturnHeaders["IMAGESCRIPT_WALLTIME"] = "x-imagescript-walltime";
    ReturnHeaders["IMAGESCRIPT_MEMORY"] = "x-imagescript-memory";
    ReturnHeaders["RATELIMIT_RESET"] = "x-rate-limit-reset";
    ReturnHeaders["RATELIMIT_LIMIT"] = "x-rate-limit-limit";
    ReturnHeaders["RATELIMIT_REMAINING"] = "x-rate-limit-remaining";
    ReturnHeaders["CONTENT_TYPE"] = "content-type";
})(ReturnHeaders = exports.ReturnHeaders || (exports.ReturnHeaders = {}));
var ContentTypes;
(function (ContentTypes) {
    ContentTypes["PNG"] = "image/png";
    ContentTypes["GIF"] = "image/gif";
})(ContentTypes = exports.ContentTypes || (exports.ContentTypes = {}));
var ImageTypes;
(function (ImageTypes) {
    ImageTypes["PNG"] = "png";
    ImageTypes["GIF"] = "gif";
})(ImageTypes = exports.ImageTypes || (exports.ImageTypes = {}));
